File: 205-isomorphic-strings/isomorphic-strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) {
            return false;
        }

        unordered_map<char, char> mapST; // Mapping from characters in s to t
        unordered_map<char, char> mapTS; // Mapping from characters in t to s

        for (int i = 0; i < s.length(); i++) {
            char charS = s[i];
            char charT = t[i];

            // Check mapping from s to t
            if (mapST.find(charS) != mapST.end()) {
                if (mapST[charS] != charT) {
                    return false; // Conflict in mapping
                }
            } else {
                mapST[charS] = charT;
            }

            // Check mapping from t to s
            if (mapTS.find(charT) != mapTS.end()) {
                if (mapTS[charT] != charS) {
                    return false; // Conflict in reverse mapping
                }
            } else {
                mapTS[charT] = charS;
            }
        }

        return true; // Strings are isomorphic
    }
};

File: 205-isomorphic-strings/isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        charST = {}
        charTS = {}

        for i in range(len(s)):
            charS = s[i]
            charT = t[i]

            if charS in charST and charST[charS] != charT:
                return False
            else:
                charST[charS] = charT

            if charT in charTS and charTS[charT] != charS:
                return False
            else:
                charTS[charT] = charS
            
        return True